### PR TITLE
Graceful fail on pipeline not found - now with tests

### PR DESCRIPTION
**Pull request by montaguegabe**
Continuation of #2

### CODE REVIEW

File: src/parsagon/api.py [hunk 1]
```diff
@@ -1,17 +1,28 @@
...
+    def __init__(self, program_name):
+        self.program_name = program_name
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if issubclass(exc_type, APIException):
+            error_value = exc_value.value
+            if isinstance(error_value, dict) and error_value.get("detail") == "Not found.":
```
Comment (Sand1929) [inline src/parsagon/api.py:23]: Seems like APIException should include data about status codes, which would make this check simpler

File: src/parsagon/api.py [hunk 2]
```diff
@@ -1,17 +1,28 @@
...
-class APIException(Exception):
-    @property
-    def value(self):
-        return self.args[0]
+class RaiseProgramNotFound:
+    def __init__(self, program_name):
+        self.program_name = program_name
+
+    def __enter__(self):
+        return self
```
Comment (Sand1929) [inline src/parsagon/api.py:18]: why does this return self?

**Review by Sand1929 — CHANGES_REQUESTED** [review]

File: src/parsagon/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+class ParsagonException(Exception):
+    """Base class for all exceptions in Parsagon. The CLI will use the to_string method to print the exception to the user."""
+
+    def to_string(self, verbose):
```
[the remark this answers is not in the document]
Comment (montaguegabe) [reply src/parsagon/exceptions.py:4]: It’s not used right now, but I think in the future we might want to give different levels of info about exceptions depending on whether the command was run with -v

File: src/parsagon/api.py (hunk 1 above)
Comment (montaguegabe) [reply src/parsagon/api.py:23]: Ah yes good idea

File: src/parsagon/api.py (hunk 2 above)
Comment (montaguegabe) [reply src/parsagon/api.py:18]: __enter__ must be defined to use the context, but the return value is unused. I can make it None

File: src/parsagon/api.py (hunk 1 above)
Comment (montaguegabe) [reply src/parsagon/api.py]: Done

File: src/parsagon/api.py (hunk 2 above)
Comment (montaguegabe) [reply src/parsagon/api.py]: Done

File: src/parsagon/tests/api_mocks.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+
+def install_api_mocks(mocker, mock_options=None):
+
+    if mock_options is None:
+        mock_options = {}
+
+    for method in ["get", "post", "put", "patch", "delete"]:
+        mocker.patch(
+            "parsagon.api.httpx.%s" % method,
```
Comment (Sand1929) [inline src/parsagon/tests/api_mocks.py:84]: Are we concerned about making the mock target not editable? e.g., if we imported httpx elsewhere, or if we changed `import httpx` to `from httpx import post` or anything like that, this target will be incorrect

**Review by Sand1929 — CHANGES_REQUESTED** [review]

File: src/parsagon/tests/api_mocks.py (hunk 4 above)
Comment (montaguegabe) [reply src/parsagon/tests/api_mocks.py:89]: Yah - that's always the problem with installing mocks in general is the messiness of having to keep track of exactly how and where things are imported.  If we changed the way it's imported then we'd have to add lines that have `mocker.patch("parsagon.api.post", ...)`, or else if httpx is used elsewhere we have to add those places too. I'm not sure of a way around this.  I can add a comment though that says something like `# NOTE: Remember to update this if we add other calling points to the backend, or if we change the way httpx method functions are imported.`

**Comment by montaguegabe** [comment]
Rebased

**Comment by montaguegabe** [comment]
Just added new way of accessing API key and API base to prevent them from being used with pytest

**Review by Sand1929 — APPROVED** [review]

**Pull request merged**